Chapter 7 Carlos corrections (#141)
chhoumann/p7: report/sections/TestRunner/chapter.tex

\chapter{Test Runner} \label{chap:TestRunner}

> When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and send back the result as fast as possible.

Author: ```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests on the submitted code and make the result available as fast as possible.
```

In order to run the tests, we run the GHC interpreter on the test code, which, in turns, needs access to the exercise submission code.
The GHC interpreter can be run from the command line with arguments specifying the path Haskell file to run.

> So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to separate files.

Author: ```suggestion
So to run the tests, the Test Runner writes the code submitted by the user and the accompanying test code to separate files.
```

However, if multiple users were to post a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
To address this, we generate a new directory with a unique name for each submission.
Then, the user submission is written to two  \texttt{.hs} files inside this directory - a \texttt{code.hs} file containing the exercise submission, and a \texttt{test.hs} file containing the corresponding tests.
This way, we can pass the path to this isolated directory to the GHC interpreter process as an argument.
Once the tests have been run, the directory and its contents are deleted again.
The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.

Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
Processing each request sequentially results in slow response times, which may also lead to connection timeouts.

> This was solved by allowing multiple instances of the GCH interpreter process to run at the same time.

Author: ```suggestion
We solved this the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```

\chapter{Test Runner} \label{chap:TestRunner}

> When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and send back the result as fast as possible.

Reviewer: ```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and make the result available as fast as possible.
```
We don't send the results back without pollling, and that violates the goal criteria otherwise.

Author: ```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests on the submitted code and make the result available as fast as possible.
```

> In order to run the tests, we run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.

Reviewer: ```suggestion
In order to run the tests, we run the GHC interpreter on the test code, which, in turns, needs access to the exercise submission code.
```

The GHC interpreter can be run from the command line with arguments specifying the path Haskell file to run.

> In other words, the Test Runner writes the user submission and corresponding test code to a file.

Reviewer: ```suggestion
So to run the tests, the Test Runner writes the user's code and corresponding exercise test code to a file.
```
This feels clearer. Removing `needs` from the sentence changed its meaning, which IMO made it a bit wonky.

And should we clarify that it's actually writing to separate files?

Author: ```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to a file.
```

Reviewer: ```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to files.
```
I don't think `a file each` is great. What do you think of this formulation?
or `separate files`?

Author: ```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to separate files.
```

However, if multiple users were to post a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
To address this, we generate a new directory with a unique name for each submission.
Then, the user submission is written to two  \texttt{.hs} files inside this directory - a \texttt{code.hs} file containing the exercise submission, and a \texttt{test.hs} file containing the corresponding tests.
This way, we can pass the path to this isolated directory to the GHC interpreter process as an argument.
Once the tests have been run, the directory and its contents are deleted again.
The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.

> Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.

Reviewer: ```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when users submit solutions in quick succession.
```

Author: The idea with "multiple" was to say that it was many users at the same time. Maybe we should just change "multiple" to many instead of removing it?

Reviewer: So it is to emphasize that it is _many_ users, not just some number of users? If so, sure, we can change multiple to another word.

Author: ```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when many multiple users submit solutions in quick succession.
```

Processing each request sequentially results in slow response times, which may also lead to connection timeouts.

> In order to address this,

Reviewer: unfinished sentence

Author: Removed it.

> This was solved by allowing multiple instances of the GCH interpreter process to run at the same time.

Reviewer: ```suggestion
We solved the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```
Clarifying which problem was solved.

Author: ```suggestion
We solved this the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```

> However, despite this being faster for a small number of clients, this approach in itself is not sufficient:

Reviewer: ```suggestion
However, despite this being faster for a small number of clients, this approach in itself is not sufficient.
```

If multiple clients post a code submission at the same time, the Test Runner will start a new process for each request.
Given a large enough number of requests, this would crash the Test Runner web server.
Therefore, we needed a way to limit the number of running GHC interpreter processes while at the same time storing subsequent requests for future processing.

> To solve the aforementioned problem, we decided to implement a queue system.

Reviewer: ```suggestion
To prevent crashes, we decided to implement a queue system.
```
In just a few lines, we mention two problems and make references to solutions. Might be better to specify which problem we're solving, like so?

Reviewer 2: Maaske udvide saetningen. "implement a queue system enabling xxxx"

Author: ```suggestion
To address the aforementioned problem where the system crashes under heavy use, we decided to implement a queue system.
```
Clarify that the system doesn't "just crash", but only does so under heavy use.

The idea is to enqueue requests for future processing if the number of currently running GHC interpreter processes reaches an upper limit.
This limit is defined as an environmental constant in the program.

> However, this introduces a new problem.
> The server can no longer directly respond to a POST request with the Test Runner results because the request may be handled at a later point in time.
> This means that if the request were enqueued for long enough, it would eventually time out.

Reviewer: ```suggestion
However, this introduced a new problem.
The server could no longer respond directly to a POST request with the Test Runner results because the request may be handled at a later point in time.
This means that if the request were enqueued for long enough, it would eventually time out.
```
tense
